File: 144. Binary Tree Preorder Traversal.java
class Solution {

    private List<Integer> result = new ArrayList<>();

    private void preorder(TreeNode root) {

        if (root == null) {
            return;
        }

        result.add(root.val);

        preorder(root.left);
        preorder(root.right);
    }

    public List<Integer> preorderTraversal(TreeNode root) {

        preorder(root);

        return result;
    }
}

File: 144. Binary Tree Preorder Traversal.cpp
class Solution {
public:
    vector<int> result;

    void preorder(TreeNode* root) {

        // Base case
        if (root == nullptr) {
            return;
        }

        // Visit root
        result.push_back(root->val);

        // Traverse left subtree
        preorder(root->left);

        // Traverse right subtree
        preorder(root->right);
    }

    vector<int> preorderTraversal(TreeNode* root) {

        preorder(root);

        return result;
    }
};
